File: 51. N-Queens.py
class Solution:
    '''
    backtracking solution, Time O(n!) Space O(n)
    we start from row 0 and to last row, in these condition, we try each col if it can place a queen and if it can,
    we go to the next row, and we backtrack the result.
    Remember to remove the one we add!
    '''
    def solveNQueens(self, n: int) -> List[List[str]]:
        cols = [0 for i in range(n)]
        major = [0 for i in range(2*n-1)]
        minor = [0 for i in range(2*n-1)]
        queens = set()
        ans = []
        def could_place(row, col):
            return not cols[col] + major[row-col] + minor[col+row]
        
        def place_queen(row, col):
            queens.add((row,col))
            cols[col] = 1
            major[row-col] = 1
            minor[col+row] = 1
            
        def remove_queen(row, col):
            queens.remove((row,col))
            cols[col] = 0
            major[row-col] = 0
            minor[col+row] = 0
            
        def addtoans():
            solution = []
            for _, col in sorted(queens):
                solution.append('.'*col + 'Q' + '.' * (n-col-1))
            ans.append(solution)
            
        def backtracking(row=0):
            for col in range(n):
                if could_place(row, col):
                    place_queen(row, col)
                    if row + 1 == n:
                        addtoans()
                    else:
                        backtracking(row+1)
                    remove_queen(row, col)
                    
        backtracking()
        return ans

File: 51. N-Queens.java
class Solution {
    int n;
    List<List<String>> ans;
    boolean[] colSeen, diagSeen, revSeen;
    class Pair {
        int row;
        int col;
        public Pair(int row, int col) {
            this.row = row;
            this.col = col;
        }
    }
    public List<List<String>> solveNQueens(int n) {
        this.n = n;
        colSeen = new boolean[n];
        diagSeen = new boolean[2*n+1];
        revSeen = new boolean[2*n+1];
        ans = new ArrayList<>();
        generate(new ArrayList<>(), n);
        return ans;
    }
    
    private void generate(List<Pair> pos, int left) {
        if (left == 0) {
            ans.add(toBoard(pos));
            return;
        }
        int i = n-left;
        for (int j=0; j<n; j++) {
            if (colSeen[j]) {
                continue;
            }
            if (diagSeen[i-j+n-1]) {
                continue;
            }
            if (revSeen[i+j]) {
                continue;
            }
            pos.add(new Pair(i, j));
            colSeen[j] = true;
            diagSeen[i-j+n-1] = true;
            revSeen[i+j] = true;
            generate(pos, left-1);
            pos.remove(pos.size()-1);
            colSeen[j] = false;
            diagSeen[i-j+n-1] = false;
            revSeen[i+j] = false;
        }
    }
    
    private List<String> toBoard(List<Pair> pos) {
        List<String> temp = new ArrayList<>();
        StringBuilder sb = new StringBuilder();
        for (Pair p : pos) {
            int col = p.col;
            for (int i=0; i<n; i++) {
                if (i==col) {
                    sb.append('Q');
                } else {
                    sb.append('.');
                }
            }
            temp.add(sb.toString());
            sb.setLength(0);
        }
        return temp;
    }
}